release-lock-file-spec.md

```diff
@@ -1,18 +1,147 @@
+represented with the `'0x'` prefix.
+
+* Prefixed: `0xdeadbeef`
+* Unprefixed: `deadbeef`
+
+
+#### Bytecode
+
+The set of EVM instructions as produced by a compiler.  Unless otherwise
+specified this should be assumed to be hexidecimal encoded and prefixed with a
+`'0x'`.
+
+#### Unlinked Bytecode
+
+Unlinked bytecode is the hexidecimal representation of a contract's EVM
+instructions which contains placeholders which are referred to as *link
+references*.
+
+* Unlinked Bytecode: `606060405260e0600073__MathLib_______________________________634d536f`
+
+
+#### Linked Bytecode
+
+Linked bytecode is the hexidecimal representation of a contract's EVM instructions which has hadd all *link references* replaced with the desired *link values*
+
+* linked Bytecode: `606060405260e06000736fe36000604051602001526040518160e060020a634d536f`
+
+
+#### Link Reference
+
+A placeholder within the hexidecimal representation of a contract's EVM instructions.
+
+`606060405260e0600073__MathLib_______________________________634d536f` the
+
+In the bytecode
+`606060405260e0600073__MathLib_______________________________634d536f`, the
+substring `__MathLib_______________________________` is a link reference.
+
+
+#### Link Value
+
+A link value is the value which can be inserted in place of a link reference.
+
+
+#### Linking
+
+The act of replacing *link references* within some bytecode with *link values*.
+
+
+#### Contract Type
+
+This term is used to refer to a specific contract or library in the package
+source.
+
+
+#### Contract Name
+
+The name found in the source code which defines a specific *contract type*.
+These names **must** conform to the regular expression
+`[a-zA-Z][-a-zA-Z0-9_]*`.
+
+There can be multiple contracts with the same *contract name* in
+a projects source files.
+
+
+#### Contract Alias
+
+This is a name chosen to reference a specific *contract type*.  Contract
+aliases must be unique within a single release lockfile.  
+
+In the case where the *contract name* is unique across a project's source files
+the *contract name* and *contract alias* **must** either be the same *or* use
+the alias naming scheme for *contract names* which are not unique.
+
+In the case where there are multiple contracts with the same *contract name* in
+a project's source files the *contract alias* **must** use the naming scheme
+`<contract-name>[<identifier>]` where `<contract-name>` is the *contract name*
+and `<identifier>` is a value conforming to the regular expression
+`[-a-zA-Z0-9_]`.
+
+Package managers **should** use the unprefixed hexidecimal representation of
+the Keccak (SHA-3) hash of the *contract type* bytecode in it's binary
+representation as the `identifier` value.
+
+In the case that the *contract types* with name collisions have the same
+bytecode such as two abstract contracts, package managers **should** fall back
+to the Keccak (SHA-3) hash of the source file in which each *contract type* is
+defined.
```

Comment:
Correct. 

It might be useful to mention that abstract contracts shouldn't be placed in the `contract_types` object, just to remove all possibility of conflicts.[1]

------------

1. The other possibility of conflict that I've thought of is the case where two files have the exact same bytecode and the exact same contract name. This is in effect a direct copy of a source file, but perhaps places in a different directory within the source. This would be bad coding practice, generally, but shouldn't be seen as an actual conflict as those contracts would be verified as the same contract.